### DISCUSSION TITLE
Setting up DisplayMemberBinding and SelectedValueBinding from code

### DISCUSSION THREAD

**Discussion by JosefNemec** [AvaloniaUI/Avalonia#14355] [Q&A] [answered]
Hi, I'm trying create attached property that I could use to automatically setup ComboBox (SelectingItemsControl in general) as selector of enum values. I got it working expect the part where I need to set up `DisplayMemberBinding` and `SelectedValueBinding`.

Setting this from xaml works:

```
DisplayMemberBinding="{Binding (p:EnumSelector).DisplayText}"
SelectedValueBinding="{Binding (p:EnumSelector).Value}"
```

but I'm struggling to figure out how to make this from code. This is my property implementation:

```csharp
private static void HandleEnumSourceChanged(SelectingItemsControl element, AvaloniaPropertyChangedEventArgs args)
{
    var options = new List<EnumSelector>();
    foreach (Enum val in Enum.GetValues((Type)args.NewValue))
    {
        options.Add(new EnumSelector(val, val.GetDescription()));
    }

    element.ItemsSource = options;
    element.Bind(SelectingItemsControl.DisplayMemberBindingProperty, new Binding(???));
    element.Bind(SelectingItemsControl.SelectedValueBindingProperty, new Binding(???));
}
```

`EnumSelector` is simple class that holds my enum value and text to be displayed for it (because I'm localizing it):

```csharp
public class EnumSelector
{
    public Enum Value { get; set; }
    public string DisplayText { get; set; }

    public EnumSelector(Enum value, string displayText)
    {
        Value = value;
        DisplayText = displayText;
    }
}
```

Basically the issue is how to set up those last two Bind calls to `Value` and `DisplayText` members? I found some information about LINQ expression bindings but that info points to binding expression classes that are internal.

**Comment by jp2masa** [comment]
I think the best way is to build a compiled binding:

https://github.com/AvaloniaUI/Avalonia/blob/b3a8ba25cf3e6f43fe89b8df91eb339bc361e482/src/Markup/Avalonia.Markup.Xaml/MarkupExtensions/CompiledBindings/CompiledBindingPath.cs#L91

https://github.com/AvaloniaUI/Avalonia/blob/b3a8ba25cf3e6f43fe89b8df91eb339bc361e482/src/Markup/Avalonia.Markup.Xaml/MarkupExtensions/CompiledBindingExtension.cs#L16

**Reply by JosefNemec** [reply]
Can't find any documentation for this so this is what I tried:

```csharp
var cb = new CompiledBindingExtension(
    new CompiledBindingPathBuilder().
        TypeCast<EnumSelector>().
        Property(
            new ClrPropertyInfo(
                nameof(EnumSelector.DisplayText),
                obj => ((EnumSelector)obj).DisplayText,
                (obj, val) => ((EnumSelector)obj).DisplayText = val as string ?? string.Empty,
                typeof(string)),
                PropertyInfoAccessorFactory.CreateInpcPropertyAccessor
            ).Build());
element.Bind(SelectingItemsControl.DisplayMemberBindingProperty, cb);
```

Doesn't work, no exceptions. Any ideas?

**Comment by gehongyan** [comment, marked as answer]
> Can't find any documentation for this so this is what I tried:
> 
> ```cs
> var cb = new CompiledBindingExtension(
>     new CompiledBindingPathBuilder().
>         TypeCast<EnumSelector>().
>         Property(
>             new ClrPropertyInfo(
>                 nameof(EnumSelector.DisplayText),
>                 obj => ((EnumSelector)obj).DisplayText,
>                 (obj, val) => ((EnumSelector)obj).DisplayText = val as string ?? string.Empty,
>                 typeof(string)),
>                 PropertyInfoAccessorFactory.CreateInpcPropertyAccessor
>             ).Build());
> element.Bind(SelectingItemsControl.DisplayMemberBindingProperty, cb);
> ```
> 
> Doesn't work, no exceptions. Any ideas?

Your creation of the `CompiledBindingExtension` looks correct, but it's important to note that both the `DisplayMemberBinding` and `SelectedValueBinding` properties are of type `IBinding?`. They expect a binding expression rather than an instantiated binding value. 

https://github.com/AvaloniaUI/Avalonia/blob/b3a8ba25cf3e6f43fe89b8df91eb339bc361e482/src/Avalonia.Controls/ItemsControl.cs#L73-L82

https://github.com/AvaloniaUI/Avalonia/blob/b3a8ba25cf3e6f43fe89b8df91eb339bc361e482/src/Avalonia.Controls/Primitives/SelectingItemsControl.cs#L246-L256

Therefore, you need to set your compiled binding instance by assignment to these properties, rather than binding your `CompiledBindingExtension` object to the `IBinding` properties again.

```csharp
// All of these should work. Choose one you prefer.
element.DisplayMemberBinding = cb;
element.SetValue(ItemsControl.DisplayMemberBindingProperty, cb);
element[ItemsControl.DisplayMemberBindingProperty] = cb;
```

![image](https://github.com/AvaloniaUI/Avalonia/assets/21241496/118ea190-b86b-42f7-9da2-77a919060be8)

**Reply by JosefNemec** [reply]
Works great, thanks a lot!
